modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala

```diff
@@ -20,7 +23,48 @@ object events {
   }
 
   object SeqexecModelUpdate {
-    implicit val equal: Equal[SeqexecModelUpdate] = Equal.equalA
+    import SeqexecEvent._
+    implicit val equal: Eq[SeqexecModelUpdate] =
+      Eq[Either[SequenceStart,
```

Comment:
This is weird. changing this to `Eq.by` produces a:
```
[error] Caused by: scala.UninitializedFieldError: Uninitialized field: /Users/cquiroz/Projects/ocs3/modules/edu.gemini.seqexec.model/src/main/scala/edu/gemini/seqexec/model/events.scala: 27
```

When I run the tests

Reply:
Curious, that means it's calling itself recursively.

Reply:
Yes. I'm pulling my hair to find out the reason, but I think I'm going to give up and do `Eq.instance`

Reply:
Yes they do run. My new instance is this one:
```
    implicit val equal: Eq[SeqexecEvent] =
      Eq.by {
        case s: ConnectionOpenEvent    => s
        case s: SequenceStart          => s
        case s: StepExecuted           => s
        case s: FileIdStepExecuted     => s
        case s: SequenceCompleted      => s
        case s: SequenceLoaded         => s
        case s: SequenceUnloaded       => s
        case s: StepBreakpointChanged  => s
        case s: OperatorUpdated        => s
        case s: ObserverUpdated        => s
        case s: ConditionsUpdated      => s
        case s: StepSkipMarkChanged    => s
        case s: SequencePauseRequested => s
        case s: SequencePauseCanceled  => s
        case s: SequenceRefreshed      => s
        case s: ActionStopRequested    => s
        case s: ResourcesBusy          => s
        case s: SequenceUpdated        => s
        case s: SequencePaused         => s
        case s: ExposurePaused         => s
        case s: SequenceError          => s
        case s: NewLogMessage          => s
        case s: ServerLogMessage       => s
        case s: NullEvent.type         => s
      }
```

Reply:
That one produces an `UnitializedError`. I'm perhaps just missing a detail, but i suspect in this case `by` is pulling a bunch of `Eq` instances per each subclass and one or more of them are not yet initialized

Reply:
Oh, the LUB of all those subtypes is `SeqexecEvent` so it's looking itself up recursively.

Reply:
What's `LUB`?

Reply:
Any ideas on how to workaround this? `Eq.instance` perhaps?

Reply:
I think `.by` nested `Either` is the right way to do it, or introduce [kittens](https://github.com/milessabin/kittens) and the derivation can be done automatically. It might make sense to start using kittens for these mechanical instances.

```scala
implicit val equal: Eq[SeqexecModelUpdate] =
  cats.derived.semi.eq
```

Reply:
`kittens` sounds good in this context

Reply:
I just tried it. Seems to work.

Reply:
perhaps I can do `Eq.instance` and rely on the `Eq` instances of subclasses members. I'm a bit wary of using macros as in kittens. but if that doesn't work autoderivation could be used

Reply:
Doing the cases pairwise with `Eq.instance` will work but you will need a catch-all case that will defeat exhaustiveness checking. But as long as you have a good `Arbitrary` and check the `Eq` laws I think this is fine.

Reply:
nice `Eq.instance` doesn't break and all test are passing